gtfs-validator/main/src/main/java/org/mobilitydata/gtfsvalidator/validator/StopTooFarFromTripShapeValidator.java

```diff
@@ -0,0 +1,116 @@
+
+import static org.locationtech.spatial4j.context.SpatialContext.GEO;
+
+/**
+ * Validates that a stop is within a distance threshold for a trip shape.
+ * <p>
+ * Generated notices:
+ * * StopTooFarFromTripShapeNotice
+ */
+@GtfsValidator
+public class StopTooFarFromTripShapeValidator extends FileValidator {
+    @Inject GtfsStopTimeTableContainer stopTimeTable;
+
+    @Inject GtfsTripTableContainer tripTable;
+
+    @Inject GtfsShapeTableContainer shapeTable;
+
+    @Inject GtfsStopTableContainer stopTable;
+
+    // Spatial operation buffer values
+    public static final double TRIP_BUFFER_METERS = 100;
+    private static final double TRIP_BUFFER_DEGREES = DistanceUtils.KM_TO_DEG * (TRIP_BUFFER_METERS / 1000.0d);
+
+    @Override
+    public void validate(NoticeContainer noticeContainer) {
+        // Cache for previously tested shape_id and stop_id pairs - no need to test them more than once.
+        final Set<String> testedCache = new HashSet<>();
+        // The maximum acceptance distance difference from the stop to the trip shape.
+        final double distanceThreshold = 1000.0;
+        // Go through the pair of tripId and the corresponding stop time list one by one.
+        for (Map.Entry<String, List<GtfsStopTime>> tripIdStopTimeListEntry : Multimaps.asMap(stopTimeTable.byTripIdMap()).entrySet()) {
+            final String tripId = tripIdStopTimeListEntry.getKey();
+            final List<GtfsStopTime> stopTimeList = tripIdStopTimeListEntry.getValue();
+            final GtfsTrip trip = tripTable.byTripId(tripId);
+            if (trip == null || stopTimeList == null || stopTimeList.isEmpty() || trip.shapeId() == null) {
+                continue;
+            }
+            final List<GtfsShape> shapeList = shapeTable.byShapeId(trip.shapeId());
+
+            // Create a polyline from the GTFS shapes data.
+            ShapeFactory.LineStringBuilder lineBuilder = getShapeFactory().lineString();
+            for (GtfsShape shape : shapeList) {
+                lineBuilder.pointXY(shape.shapePtLon(), shape.shapePtLat());
+            }
+            Shape shapeLine = lineBuilder.build();
+            // Create the buffered version of the trip as a polygon.
+            Shape shapeBuffer = shapeLine.getBuffered(TRIP_BUFFER_DEGREES, shapeLine.getContext());
+
+            // Check if each stop is within the buffer polygon.
+            for (GtfsStopTime stopTime : stopTimeList) {
+                GtfsStop stop = stopTable.byStopId(stopTime.stopId());
+                if (stop == null || shapeList == null || shapeList.isEmpty()) {
+                    continue;
+                }
+                if (!(stop.locationType() == GtfsLocationType.STOP) && !(stop.locationType() == GtfsLocationType.BOARDING_AREA)) {
+                    // This rule only applies to stops of location_type "STOP" or "BOARDING_AREA".
+                    continue;
+                }
+                if (testedCache.contains(trip.shapeId() + stop.stopId())) {
```

Comment:
All Ids are strings. We use "trip.shapeId() + stop.stopId()" to combine a unique String key which makes the trip shape and stop pair only need to be checked once.
I have added comments for "trip.shapeId() + stop.stopId()" related parts in codes.

Reply:
What is the guarantee that trip.shapeId() + stop.stopId() is an appropriate unique identifier.

Is it possible to have
trip1.shapeId = 'abcd' ; stop1.stopId = 'ef'
trip2.shapeId = 'abc' ;  stop2.stopId = 'def'
?

If you used AutoValue for this, you could enjoy a helpful hash function.

Reply:
Thank you for the example! Yes, it could happen and the validator will miss some pairs.
I currently plan to use HashMap for the records rather than the Set of String - I think it is an easy way that can record pairs correctly and is also easy to check the duplicated pairs later.
BTW, I want to ask, if we use AutoValue here, does that mean we need to add the "AutoValue" annotations for the corresponding types (i.e. GtfsTrip, GtfsStop) in the original files, or we need to create new corresponding types?

Reply:
Done!